src/lib/Output/Generator/Xml.php

```diff
@@ -4,264 +4,129 @@
+    public const string OUTER_ELEMENT = 'outer_element';
 
-    /**
-     * Keeps track if the document received some content.
-     *
-     * @var bool
-     */
-    protected $isEmpty = true;
-
-    /**
-     * Enables developer to modify REST response media type prefix.
-     *
-     * @var string
-     */
-    protected $vendor;
-
-    /**
-     * @param \Ibexa\Rest\Output\Generator\Xml\FieldTypeHashGenerator $hashGenerator
-     * @param string $vendor
-     */
-    public function __construct(Xml\FieldTypeHashGenerator $hashGenerator, $vendor = 'vnd.ibexa.api')
-    {
-        $this->hashGenerator = $hashGenerator;
+    public function __construct(
+        Xml\FieldTypeHashGenerator $fieldTypeHashGenerator,
+        string $vendor = 'vnd.ibexa.api',
+    ) {
+        $this->fieldTypeHashGenerator = $fieldTypeHashGenerator;
         $this->vendor = $vendor;
     }
 
-    /**
-     * Start document.
-     *
-     * @param mixed $data
-     */
-    public function startDocument($data)
-    {
-        $this->checkStartDocument($data);
-
-        $this->isEmpty = true;
-
-        $this->xmlWriter = new \XMLWriter();
-        $this->xmlWriter->openMemory();
-        $this->xmlWriter->setIndent($this->formatOutput);
-        $this->xmlWriter->startDocument('1.0', 'UTF-8');
-    }
-
-    /**
-     * Returns if the document is empty or already contains data.
-     *
-     * @return bool
-     */
-    public function isEmpty()
-    {
-        return $this->isEmpty;
-    }
-
-    /**
-     * End document.
-     *
-     * Returns the generated document as a string.
-     *
-     * @param mixed $data
-     *
-     * @return string
-     */
-    public function endDocument($data)
+    #[\Override]
+    public function getMediaType(string $name): string
     {
-        $this->checkEndDocument($data);
-
-        $this->xmlWriter->endDocument();
-
-        return $this->xmlWriter->outputMemory();
+        return $this->generateMediaTypeWithVendor($name, 'xml', $this->vendor);
     }
 
-    /**
-     * Start object element.
-     *
-     * @param string $name
-     * @param string $mediaTypeName
-     */
-    public function startObjectElement($name, $mediaTypeName = null)
+    #[\Override]
+    public function startList(string $name): void
     {
-        $this->checkStartObjectElement($name);
+        $this->checkStartList($name);
 
         $this->isEmpty = false;
 
-        $mediaTypeName = $mediaTypeName ?: $name;
+        $array = new Data\ArrayList($name, $this->json);
 
-        $this->xmlWriter->startElement($name);
-
-        $this->startAttribute('media-type', $this->getMediaType($mediaTypeName));
-        $this->endAttribute('media-type');
-    }
-
-    /**
-     * End object element.
-     *
-     * @param string $name
-     */
-    public function endObjectElement($name)
-    {
-        $this->checkEndObjectElement($name);
-
-        $this->xmlWriter->endElement();
+        $this->json->$name = $array;
+        $this->json = $array;
```

Comment:
The first line doesn't make sense, as it's overwritten in the second line.

Reply:
Actually, it has sense.

Two things happen:
1. We are assigning a property (with an object `[1]`) to an object `[2]`.
2. We are replacing **a reference**. We're placing object `[1]` where object `[2]` was (changing current pointer).

Object `[2]` contains a reference to object `[1]`. Whenever we call `::endXXX()` methods, we replace current state with parent back again.